File: term4/prog eng/lab1/Matrix.cpp
//---------------------------------------------------------------------------

#pragma hdrstop

#include "Matrix.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

Matrix::Matrix(int rows, int cols) : rows(rows), cols(cols), data(rows, std::vector<int>(cols, 0)) {}

Matrix::Matrix(const std::vector<std::vector<int>>& data) : data(data), rows(data.size()), cols(data[0].size()) {}

int Matrix::getRows() const { return rows; }
int Matrix::getCols() const { return cols; }
std::vector<std::vector<int>> Matrix::getData() const { return data; }

int Matrix::getColNullCount(int col) const {
    int count = 0;
    for (int i = 0; i < rows; i++) {
        if (data[i][col] == 0) {
            count++;
        }
    }
    return count;
}

int Matrix::getNullCount() const {
    int count = 0;
    for (int i = 0; i < cols; i++) {
        count += getColNullCount(i);
    }
    return count;
}

int Matrix::getMaxNullCol() const {
    int col_index = 0;
    int col_val = getColNullCount(0);
    for (int i = 1; i < cols; i++) {
        int val = getColNullCount(i);
        if (col_val < val) {
            col_val = val;
            col_index = i;
        }
    }
    return col_index;
}

bool Matrix::isNullOrdered() const {
    int prevVal = getColNullCount(0);
    for (int i = 1; i < cols; i++) {
        int val = getColNullCount(i);
        if (val < prevVal) {
            return false;
        }
        prevVal = val;
    }
    return true;
}

void Matrix::display() const {
    for (const auto& row : data) {
        for (int val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
}


// Lab 2

int Matrix::getRowSum(int i) const {
	int result = 0;
	if (i < 0 || i >= data.size()) {
		return 0;
	}
	for (int j = 0; j < data[i].size(); j++) {
		result += data[i][j];
	}
	return result;
}

int Matrix::getRowNulls(int i) const {
	int result = 0;

	if (i < 0 || i >= data.size()) {
		return 0;
	}

	for (int j = 0; j < data[i].size(); j++) {
		if (data[i][j] == 0) {
			result++;
		}
	}

	return result;
}
float Matrix::getCoff() const {
	int evenSum = 0;
	int oddSum = 0;

	for (int i = 0; i < data.size(); i++) {
		int rowSum = getRowSum(i);
		if (i % 2 == 0) {
			evenSum += rowSum;
		} else {
			oddSum += rowSum;
		}
	}
	if (oddSum == 0) {
		return 0; ;
	}

	return static_cast<float>(evenSum) / oddSum;
}

int Matrix::getEvenRowsNulls() const {
	int result = 0;

	for (int i = 0; i < data.size(); i++) {
		if (i % 2 == 0) {
			result += getRowNulls(i);
		}
	}

	return result;
}

File: term4/prog eng/lab5/Matrix.pas
unit Matrix;

interface

type
  TMatrixData = array of array of Integer;

  TMatrix = class
  private
    FRows: Integer;
    FCols: Integer;
    FData: TMatrixData;
  public
    constructor Create(Rows, Cols: Integer); overload;
    constructor Create(const Data: TMatrixData); overload;

    function GetRows: Integer;
    function GetCols: Integer;
    function GetData: TMatrixData;

    function GetColNullCount(Col: Integer): Integer;
    function GetNullCount: Integer;
    function GetMaxNullCol: Integer;
    function IsNullOrdered: Boolean;
    procedure Display;

    // Lab 2
    function GetRowSum(I: Integer): Integer;
    function GetRowNulls(I: Integer): Integer;
    function GetCoff: Single;
    function GetEvenRowsNulls: Integer;

    procedure InitializeMatrix;

    property Rows: Integer read GetRows;
    property Cols: Integer read GetCols;
    property Data: TMatrixData read GetData;
  end;

var
  AMatrix: TMatrix;
  matrixSize: Integer;

implementation

uses
  SysUtils;

constructor TMatrix.Create(Rows, Cols: Integer);
var
  I: Integer;
begin
  FRows := Rows;
  FCols := Cols;
  SetLength(FData, Rows, Cols);
end;

constructor TMatrix.Create(const Data: TMatrixData);
begin
  FData := Data;
  FRows := Length(Data);
  if FRows > 0 then
    FCols := Length(Data[0])
  else
    FCols := 0;
end;

function TMatrix.GetRows: Integer;
begin
  Result := FRows;
end;

function TMatrix.GetCols: Integer;
begin
  Result := FCols;
end;

function TMatrix.GetData: TMatrixData;
begin
  Result := FData;
end;

function TMatrix.GetColNullCount(Col: Integer): Integer;
var
  I: Integer;
begin
  Result := 0;
  for I := 0 to FRows - 1 do
    if FData[I, Col] = 0 then
      Inc(Result);
end;

function TMatrix.GetNullCount: Integer;
var
  I: Integer;
begin
  Result := 0;
  for I := 0 to FCols - 1 do
    Result := Result + GetColNullCount(I);
end;

function TMatrix.GetMaxNullCol: Integer;
var
  I, ColVal, MaxVal: Integer;
begin
  Result := 0;
  MaxVal := GetColNullCount(0);

  for I := 1 to FCols - 1 do
  begin
    ColVal := GetColNullCount(I);
    if MaxVal < ColVal then
    begin
      MaxVal := ColVal;
      Result := I;
    end;
  end;
end;

function TMatrix.IsNullOrdered: Boolean;
var
  I, PrevVal, Val: Integer;
begin
  PrevVal := GetColNullCount(0);

  for I := 1 to FCols - 1 do
  begin
    Val := GetColNullCount(I);
    if Val < PrevVal then
    begin
      Result := False;
      Exit;
    end;
    PrevVal := Val;
  end;
  Result := True;
end;

procedure TMatrix.Display;
var
  I, J: Integer;
begin
  for I := 0 to FRows - 1 do
  begin
    for J := 0 to FCols - 1 do
      Write(FData[I, J], ' ');
    Writeln;
  end;
end;

procedure TMatrix.InitializeMatrix;
var
  i, j: Integer;
begin
  if Assigned(AMatrix) then
    AMatrix.Free;
  AMatrix := TMatrix.Create(matrixSize, matrixSize);
  for i := 0 to matrixSize - 1 do
    for j := 0 to matrixSize - 1 do
      AMatrix.Data[i, j] := Random(10); // Заполняем случайными числами от 0 до 9
end;

// Lab 2

function TMatrix.GetRowSum(I: Integer): Integer;
var
  J: Integer;
begin
  Result := 0;
  if (I < 0) or (I >= FRows) then
    Exit;

  for J := 0 to FCols - 1 do
    Result := Result + FData[I, J];
end;

function TMatrix.GetRowNulls(I: Integer): Integer;
var
  J: Integer;
begin
  Result := 0;
  if (I < 0) or (I >= FRows) then
    Exit;

  for J := 0 to FCols - 1 do
    if FData[I, J] = 0 then
      Inc(Result);
end;

function TMatrix.GetCoff: Single;
var
  I, EvenSum, OddSum: Integer;
begin
  EvenSum := 0;
  OddSum := 0;

  for I := 0 to FRows - 1 do
  begin
    if I mod 2 = 0 then
      EvenSum := EvenSum + GetRowSum(I)
    else
      OddSum := OddSum + GetRowSum(I);
  end;

  if OddSum = 0 then
    Result := 0
  else
    Result := EvenSum / OddSum;
end;

function TMatrix.GetEvenRowsNulls: Integer;
var
  I: Integer;
begin
  Result := 0;
  for I := 0 to FRows - 1 do
    if I mod 2 = 0 then
      Result := Result + GetRowNulls(I);
end;
end.
